File: toolargetool/src/main/java/com/gu/toolargetool/ActivitySavedStateLogger.java
package com.gu.toolargetool;

import android.app.Activity;
import android.os.Bundle;

import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentActivity;

/**
 * {@link android.app.Application.ActivityLifecycleCallbacks} implementation that logs information
 * about the saved state of Activities.
 */
public class ActivitySavedStateLogger extends EmptyActivityLifecycleCallbacks {
    @Nullable private final FragmentSavedStateLogger fragmentLogger;
    @NonNull private final Map<Activity, Bundle> savedStates = new HashMap<>();
    @NonNull private final TooLargeLoggerCallback activityCallback;

    ActivitySavedStateLogger(@NonNull TooLargeLoggerCallback activityCallback, @Nullable TooLargeLoggerCallback fragmentCallback) {
        this.fragmentLogger = fragmentCallback != null ? new FragmentSavedStateLogger(fragmentCallback) : null;
        this.activityCallback = activityCallback;
    }


    @Override
    public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
        if (activity instanceof FragmentActivity && fragmentLogger != null) {
            ((FragmentActivity) activity)
                    .getSupportFragmentManager()
                    .registerFragmentLifecycleCallbacks(fragmentLogger, true);
        }
    }

    @Override
    public void onActivityDestroyed(Activity activity) {
        if (activity instanceof FragmentActivity && fragmentLogger != null) {
            ((FragmentActivity) activity)
                    .getSupportFragmentManager()
                    .unregisterFragmentLifecycleCallbacks(fragmentLogger);
        }
    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
        savedStates.put(activity, outState);
    }

    @Override
    public void onActivityStopped(Activity activity) {
        Bundle savedState = savedStates.remove(activity);
        if (savedState != null) {
            activityCallback.log(activity.getClass().getSimpleName() + ".onSaveInstanceState wrote: " + TooLargeTool.bundleBreakdown(savedState));
        }
    }
}

File: toolargetool/src/main/java/com/gu/toolargetool/ActivitySavedStateLogger.kt
package com.gu.toolargetool

import android.app.Activity
import android.app.Application
import android.os.Bundle
import androidx.fragment.app.FragmentActivity
import java.util.*

/**
 * [android.app.Application.ActivityLifecycleCallbacks] implementation that logs information
 * about the saved state of Activities.
 */
internal class ActivitySavedStateLogger(
        private val activityCallback: ActivityCallback,
        fragmentCallback: FragmentCallback?
) : Application.ActivityLifecycleCallbacks {

    private val fragmentLogger: FragmentSavedStateLogger? = if (fragmentCallback != null) FragmentSavedStateLogger(fragmentCallback) else null
    private val savedStates = HashMap<Activity, Bundle>()


    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle) {
        if (activity is FragmentActivity && fragmentLogger != null) {
            activity.supportFragmentManager
                    .registerFragmentLifecycleCallbacks(fragmentLogger, true)
        }
    }

    override fun onActivityDestroyed(activity: Activity) {
        if (activity is FragmentActivity && fragmentLogger != null) {
            activity.supportFragmentManager
                    .unregisterFragmentLifecycleCallbacks(fragmentLogger)
        }
    }

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
        savedStates[activity] = outState
    }

    override fun onActivityStopped(activity: Activity) {
        val savedState = savedStates.remove(activity)
        savedState?.let {
            activityCallback.callback(activity, it)
        }
    }

    override fun onActivityPaused(p0: Activity?) {
        //NOP
    }

    override fun onActivityResumed(p0: Activity?) {
        //NOP
    }

    override fun onActivityStarted(p0: Activity?) {
        //NOP
    }
}
